src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,973 @@
+                                      (unwrap! (get name name-currently-owned) (err ERR_PANIC)))
+                                    (err ERR_PANIC)))))
+      ;; The sender principal must match the namespace's import principal
+      (asserts!
+        (is-eq (get namespace-import namespace-props) contract-caller)
+        (err ERR_NAMESPACE_OPERATION_UNAUTHORIZED))
+      ;; The name's namespace must not be launched
+      (asserts!
+        (is-none (get launched-at namespace-props))
+        (err ERR_NAMESPACE_ALREADY_LAUNCHED))
+      ;; The principal can register a name. Should we enable this condition in case of NAME-IMPORT
+      ;; (asserts!
+      ;;   can-sender-register-name
+      ;;   (err ERR_PRINCIPAL_ALREADY_ASSOCIATED))
+      ;; Less than 1 year must have passed since the namespace was "revealed"
+      (asserts!
+        (< block-height (+ (get revealed-at namespace-props) NAMESPACE_LAUNCHABILITY_TTL))
+        (err ERR_NAMESPACE_PREORDER_LAUNCHABILITY_EXPIRED))
+      ;; Mint the new name
+      (unwrap! 
+        (nft-mint? names (tuple (namespace namespace) (name name)) contract-caller)
+        (err ERR_NAME_UNAVAILABLE))
+      ;; Attach the new name
+      (map-set owner-name
+        ((owner contract-caller))
+        ((namespace namespace) (name name)))
+      ;; Update zonefile and props
+      (update-zonefile-and-props
+        namespace 
+        name  
+        none
+        (some block-height) ;; Set imported-at
+        none
+        zonefile-hash)
+      (ok true))))
+
+;; NAMESPACE_READY
+;; The final step of the process launches the namespace and makes the namespace available to the public. Once a namespace
+;; is launched, anyone can register a name in it if they pay the appropriate amount of cryptocurrency.
+(define-public (namespace-ready (namespace (buff 19)))
+  (let (
+      (namespace-props (unwrap!
+        (map-get? namespaces ((namespace namespace)))
+        (err ERR_NAMESPACE_NOT_FOUND)))
+      (owned-name (map-get? owner-name ((owner contract-caller)))))
+    ;; The sender principal must match the namespace's import principal
+    (asserts!
+      (is-eq (get namespace-import namespace-props) contract-caller)
+      (err ERR_NAMESPACE_OPERATION_UNAUTHORIZED))
+    ;; The name's namespace must not be launched
+    (asserts!
+      (is-none (get launched-at namespace-props))
+      (err ERR_NAMESPACE_ALREADY_LAUNCHED))
+    ;; Less than 1 year must have passed since the namespace was "revealed"
+    (asserts!
+      (< block-height (+ (get revealed-at namespace-props) NAMESPACE_LAUNCHABILITY_TTL))
+      (err ERR_NAMESPACE_PREORDER_LAUNCHABILITY_EXPIRED))
+    ;; Update eventual name-props
+    (if (is-none owned-name)
+      true
+      (let ((name (unwrap! (get name owned-name) (err ERR_PANIC))))
+        (let ((name-props (unwrap! (map-get? name-properties ((name name) (namespace namespace))) (err ERR_PANIC))))
```

Comment:
I'm not sure what's happening here?  Why is this method involving any names?

Reply:
The namespace-reveal principal can import an unbound number of names while the namespace is not launched.  There's no way to switch all of these names' `imported-at` and `registered-at` fields in Clarity, since this would constitute an unbound list iteration.

Stacks 1.0 doesn't do this either.  It simply considers a name's `imported-at` and `registered-at` fields along with its namespace's `launched-at` field to determine how old it is -- if the name has an `imported-at` value that is not `none`, then the name was "registered" at the namespace's `launched-at` block height.

Reply:
I think my understanding of imported names was incomplete. To your question, I was applying the rule "1 principal, 1 name", which sounds not desirable.
Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/58c72297359afda14cac67254ae854c93387a583